packages/data/src/components/use-select/index.js

```diff
@@ -19,6 +19,25 @@ import useAsyncMode from '../async-mode-provider/use-async-mode';
 
 const renderQueue = createQueue();
 
+function warnOnUnstableReference( a, b ) {
+	if ( ! a && ! b ) {
+		return;
+	}
+
+	const keys =
+		typeof a === 'object' && typeof b === 'object'
+			? Object.keys( a ).filter( ( k ) => a[ k ] !== b[ k ] )
+			: [];
+
+	// eslint-disable-next-line no-console
+	console.warn(
+		'The `useSelect` hook returns different values when called with the same state and parameters.\n' +
+			'This can lead to unnecessary re-renders and performance issues if not fixed.\n\n' +
+			'Non-equal value keys: %s\n\n',
+		keys.join( ', ' )
+	);
```

Comment:
The warning will be confusing when the `b` object has some field that the `a` object doesn't have. Then it will report an empty list of non-equal keys.

It can happen in practice when we write code like this:
```js
useSelect( select => {
  if ( someCondition ) {
    return { loading: false };
  }
  return { loading: false, data: select( store ).getData() };
} );
```
To fix this, we could construct a set of all fields on both objects and then filter it:
```js
const allKeys = Array.from( new Set( [ ...Object.keys( a ), ...Object.keys( b ) ] ) );
const keys = allKeys.filter( ( k ) => a[ k ] !== b[ k ] );
```
But this will fail again for objects like `{}` and `{ a: undefined }` which are not shallow equal but `keys` will be `[]`.

Reply:
> The warning will be confusing when the b object has some field that the a object doesn't have. Then it will report an empty list of non-equal keys.

It means that the `b` run happened with different state or parameters (mostly params, based on your example). That shouldn't be the case for this warning, as it only makes sense with the same state/params.

Reply:
Oh that's true, I've been overthinking it 🙂 Running the `mapSelect` callback twice with the same state shouldn't produce objects like this unless you're doing it on purpose.